pkg/informer/informer.go

```diff
@@ -643,3 +663,142 @@ func (d *DynamicDiscoverySharedInformerFactory) Unsubscribe(id string) {
+}
+
+var _ discovery.DiscoveryInterface = &DynamicDiscoverySharedInformerFactory{}
+
+func (d *DynamicDiscoverySharedInformerFactory) ServerResourcesForGroupVersion(groupVersion string) (*metav1.APIResourceList, error) {
+	d.informersLock.RLock()
+	defer d.informersLock.RUnlock()
+
+	for _, apiResourceList := range d.discoveryData {
+		if apiResourceList.GroupVersion == groupVersion {
+			return apiResourceList.DeepCopy(), nil
+		}
+	}
+
+	// ignore 403 or 404 error to be compatible with a v1.0 server.
+	if groupVersion == "v1" {
+		return &metav1.APIResourceList{GroupVersion: groupVersion}, nil
+	}
+
+	return nil, errors.NewNotFound(schema.GroupResource{Group: groupVersion}, "")
+}
+
+func (d *DynamicDiscoverySharedInformerFactory) ServerGroupsAndResources() ([]*metav1.APIGroup, []*metav1.APIResourceList, error) {
+	d.informersLock.RLock()
+	defer d.informersLock.RUnlock()
+
+	retGroups := make([]*metav1.APIGroup, len(d.discoveryData))
+	failedGroups := make(map[schema.GroupVersion]error)
+	retResourceList := make([]*metav1.APIResourceList, len(d.discoveryData))
+
+	for i, apiResourceList := range d.discoveryData {
+		gv, err := schema.ParseGroupVersion(apiResourceList.GroupVersion)
+		if err != nil {
+			failedGroups[schema.GroupVersion{Group: apiResourceList.GroupVersion}] = err
+			continue
+		}
+		retGroups[i] = &metav1.APIGroup{
+			Name: gv.Group,
+			Versions: []metav1.GroupVersionForDiscovery{
+				{
+					GroupVersion: gv.String(),
+					Version:      gv.Version,
+				},
+			},
+			PreferredVersion: metav1.GroupVersionForDiscovery{
+				GroupVersion: gv.String(),
+				Version:      gv.Version,
+			},
+		}
+
+		retResourceList[i] = apiResourceList.DeepCopy()
+	}
+
+	if len(failedGroups) > 0 {
+		return retGroups, retResourceList, &discovery.ErrGroupDiscoveryFailed{Groups: failedGroups}
+	}
+
+	return retGroups, retResourceList, nil
+}
+
+func (d *DynamicDiscoverySharedInformerFactory) ServerPreferredResources() ([]*metav1.APIResourceList, error) {
+	d.informersLock.RLock()
+	defer d.informersLock.RUnlock()
+
+	ret := make([]*metav1.APIResourceList, len(d.discoveryData))
+	for i, apiResourceList := range d.discoveryData {
+		ret[i] = apiResourceList.DeepCopy()
+	}
+
+	return ret, nil
+}
+
+func (d *DynamicDiscoverySharedInformerFactory) ServerPreferredNamespacedResources() ([]*metav1.APIResourceList, error) {
+	d.informersLock.RLock()
+	defer d.informersLock.RUnlock()
+
+	ret := make([]*metav1.APIResourceList, len(d.discoveryData))
+	for i, apiResourceList := range d.discoveryData {
+		namespacedResources := &metav1.APIResourceList{GroupVersion: apiResourceList.GroupVersion}
+		for _, resource := range apiResourceList.APIResources {
+			if resource.Namespaced {
+				namespacedResources.APIResources = append(namespacedResources.APIResources, resource)
+			}
+		}
+		ret[i] = namespacedResources
+	}
+
+	return ret, nil
+}
+
+func (d *DynamicDiscoverySharedInformerFactory) RESTClient() rest.Interface {
```

Comment:
Yikes, what is pulling this in? What interface do we need to fulfill? Can we refactor that?

Reply:
This was pulled because of the RESTMapper argument passed to the GC controller factory method:

```go
restmapper.NewDeferredDiscoveryRESTMapper(memory.NewMemCacheClient(c.dynamicDiscoverySharedInformerFactory))
```

Thinking through the whole calls sequence, it turns out it's possible to remove this and be completely "push-based", while the cached RESTMapper made sense for the polling approach.

I've addressed it at the same as optimising the discovery API implementation, so discovery and REST mapping methods share the same data structures.